File: cs50x/psets-labs/sentimental-credit/credit.py
from cs50 import get_int


def main():
    cardnumber = input("Number: ")

    n = str(cardnumber)
    m = len(n)

    if CheckSum(cardnumber):
        if n[0] == "4" and (m == 13 or m == 16):
            print("VISA\n")  # Check for Visa
        elif n[0] == "3" and (n[1] == "4" or n[1] == "7") and m == 15:
            print("AMEX\n")  # Check for AMEX
        elif n[0] == "5" and (n[1] in "12345") and m == 16:
            print("MASTERCARD\n")  # Check FOr MASTERCARD
        else:
            print("INVALID\n")
    else:
        print("INVALID\n")


# FUNCTIONS
def CheckSum(cardnumber):
    cdnum = str(
        cardnumber
    )  # Converting String to int to extract alternate digits through indices
    list_cddig = []
    cddig = cdnum[-2::-2]  # Accessing digits second to last
    for j in range(len(cddig)):
        list_cddig.append(int(cddig[j]))
    cddig1 = ""
    for i in list_cddig:  # Multiplying digits by 2
        dig = i * 2
        cddig1 += str(dig)
    cddig = int(cddig1)
    sum1 = sum_of_digits(cddig)
    cddig2 = int(cdnum[-1::-2])
    sum2 = sum_of_digits(cddig2)
    sumt = sum1 + sum2
    if sumt % 10 == 0:  # Seeing if Cardnumber is valid or not
        return True
    else:
        return False


def sum_of_digits(number):  # Function to find sum of numbers
    sum = 0
    while number > 0:
        digit = number % 10
        sum += digit
        number = number // 10
    return sum


main()

File: cs50x/psets-labs/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool CheckSum(long Cardnumber) //Defining Checksum as a function
{
    long c = Cardnumber;
    long c2 = Cardnumber; // assigning card number to another variable to perform two sets of operations on it
    int sum = 0;
    bool alt = false;

    while (c > 0) // Finding sum of alternate digits from second to last
    {
        int digit = c % 10;
        if (alt)
        {
            digit = digit * 2;
            if (digit > 9)
            {
                digit = digit % 10 + digit / 10;
            }
            sum = sum + digit;
        }
        c = c / 10;
        alt = !alt; // Make alt True and False alternately
    }

    int sum2 = 0;
    bool alt2 = true;

    while (c2 > 0) // Finding sum of alternate digits from first to second last
    {
        int digit2 = c2 % 10;
        if (alt2)
        {
            sum2 = digit2 + sum2;
        }
        c2 = c2 / 10;
        alt2 = !alt2; // Make alt True and False alternately
    }
    return (sum + sum2) % 10 == 0;
}
int main(void)
{
//Chcking if card number is valid,if so which type of card
    long Cardnumber = get_long("Number: ");
    if (CheckSum(Cardnumber))
    {
        if ((Cardnumber >= 4000000000000 && Cardnumber <= 4999999999999) ||
            (Cardnumber > 4000000000000000 && Cardnumber <= 4999999999999999))
        {
            printf("VISA\n");// Check for Visa
        }
        else if ((Cardnumber >= 340000000000000 && Cardnumber <= 349999999999999) ||
                 (Cardnumber >= 370000000000000 && Cardnumber <= 379999999999999))
        {
            printf("AMEX\n");//Check for American Express
        }
        else if (Cardnumber >= 5100000000000000 && Cardnumber <= 5599999999999999)
        {
            printf("MASTERCARD\n");//Check for Mastercard
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}